File: core/src/ru/myitschool/dcrawler/skills/targeting/RaytraceDisplayer.java
package ru.myitschool.dcrawler.skills.targeting;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.ai.AIUtils;
import ru.myitschool.dcrawler.skills.Skill;
import ru.myitschool.dcrawler.utils.AdvancedArray;

/**
 * Created by Voyager on 14.03.2018.
 */
public class RaytraceDisplayer implements TargetDisplayer {

    @Override
    public boolean targetCreation(int x, int y, Array<TilePos> array, Skill skill) {
        if (Math.abs(skill.getDoer().getTileX() - x) + Math.abs(skill.getDoer().getTileY() - y) > 1) {
            boolean obstruct = skill.isObstruct();
            AdvancedArray<Vector2> cells = AIUtils.getCellRaytrace(skill.getDoer().getTileX(), skill.getDoer().getTileY(), x, y, 0);
            cells.clip(1, cells.size - 2);
            int max = cells.size;
            Array<Integer> poses = AIUtils.getObstructorIndexes(cells);
            System.out.println("X: " + x + " Y: " + y);
            if (obstruct) {
                if (poses.size > 0) {
                    max = poses.get(0);
                }
            }
            for (int i = 0; i < max; i++) {
                Vector2 cell = cells.get(i);
                array.add(new TilePos((int) cell.x, (int) cell.y, TargetRenderer.getDefaultLineTile()));
            }
            if (poses.size > 0 && obstruct) {
                Vector2 cell = cells.get(poses.get(0));
                array.add(new TilePos((int) cell.x, (int) cell.y, TargetRenderer.getDefaultObstructionTile()));
                return false;
            }
        }
        return true;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/targeting/RaytraceDisplayer.kt
package com.wolg_vlad.dcrawler.entities.skills.targeting

import com.wolg_vlad.dcrawler.entities.ai.AIUtils
import com.wolg_vlad.dcrawler.entities.skills.Skill

/**
 * Created by Voyager on 14.03.2018.
 */
class RaytraceDisplayer : TargetDisplayer {
        override fun targetCreation(x: Int, y: Int, array: MutableList<TilePos>, skill: Skill): Boolean {
        if (Math.abs(skill.doer.tileX - x) + Math.abs(skill.doer.tileY - y) > 1) {
            val obstruct = skill.isObstruct
            val cells = AIUtils.getCellRaytrace(skill.doer.tileX, skill.doer.tileY, x, y, 0)
            cells!!.clip(1, cells.size - 2)
            var max = cells.size
            val poses = AIUtils.getObstructorIndexes(cells)
            println("X: $x Y: $y")
            if (obstruct) {
                if (poses!!.size > 0) {
                    max = poses[0]!!
                }
            }
            for (i in 0 until max) {
                val cell = cells[i]
                array.add(TilePos(cell!!.x.toInt(), cell.y.toInt(), TargetRenderer.defaultLineTile))
            }
            if (poses!!.size > 0 && obstruct) {
                val cell = cells[poses[0]!!]
                array.add(TilePos(cell!!.x.toInt(), cell.y.toInt(), TargetRenderer.defaultObstructionTile))
                return false
            }
        }
        return true
    }
}